### ISSUE TITLE
CMake build with toolchain file, binaries not copied to output folder

### ISSUE THREAD

**Issue by stekyne** [Microsoft/vcpkg#1002]
I have a large Cmake project (Csound) that I'm trying to integrate VCPKG with. It builds successfully in VS and command line. It can find it's dependencies correctly and they're included/linked with correctly but the vcpkg binaries are not copied to the output directory after a build.

I've run a small test project using the same method and the vcpkg .dll files are copied as expected. This is happening across a few developer installations so it's not specific to my machine. 

How can I diagnose this? I can't really see at which point the dependencies are copied or what that mechanism is exactly. 

Thanks

**Comment by bagong** [comment]
Old style, using cmake, this would be done with some variant of Bundle Utilities ( `fixup_bundle`, or if you use Qt `install_qt5_executable`):

https://cmake.org/cmake/help/v3.0/module/BundleUtilities.html

A utility inspects your executables and copies the required shared libraries into your application folder (on Windows).

I doubt vcpkg would interfere with that process, after all it just provides libraries, checks on build integrity, and helps your project to find the libraries. I doubt it helps for the install of your project, but I'd love to learn about that.

**Comment by stekyne** [comment]
No, vcpkg should do this. Otherwise I have to manually specify what dependent DLL I need to copy for my executable to run. 

As per the example page under the CMake example, "Note: The correct sqlite3.dll is automatically copied to the output folder when building for x86-windows. You will need to distribute this along with your application."

This does work locally for a really simple CMake example that I have so I know that this should be the case.  I just need to know how to debug this process so I can see where it's going wrong.

**Comment by tobiaskohlbau** [comment]
https://github.com/Microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L111 here an post build command gets added for every executable. It should be possible to see some output of this script in post build step within VS. Otherwise you could extract the post build step and launch an native tools command prompt and check the output.

The script which gets executed is located here: https://github.com/Microsoft/vcpkg/blob/master/scripts/buildsystems/msbuild/applocal.ps1.

**Comment by stekyne** [comment]
OK great thanks. I'll see if I can run that separately and get some good logging. I've seen that script run for each project within VS but I don't think there is any real output, just a log of the script that is running.

**Comment by ras0219-msft** [comment]
We have a few `Write-Verbose` commands inside, so if you run it directly from powershell you can add `-verbose` to get a bit more information.

**Comment by stekyne** [comment]
@bagong I understand what you mean now about Windows utility. It seems dumpbin is used to find the DLLs the exe is reliant on and copies them. 

I can see the applocal.ps1 script running for all the executable files within the project. However, some of my DLLs within my own project have dependent DLLs in vcpkg but this applocal.ps1 script doesn't run for those.

Is this a limitation of vcpkg or do I need to do something else?

**Comment by bagong** [comment]
It's nice that we both have the ear of experts here, so hopefully somebody else can give a conclusive answer, and we both learn from it. In my limited understanding the scripts above are only used to copy libraries within the vcpkg system. So this would work for you only if Csound were a part of vcpkg, which of course it isn't, and cannot be, as it is a full fledged application. So likely in order to get all required libs into your own application install, you still need additional means. Again -
 from my limited understanding - you would use the cmake bundle utilities if you need to be flexible as to the libraries that are actually needed in the install of your application: some of them can be determined easily (so you could simply copy them), but others, like system runtime libraries, or indirect dependencies, are less easy to determine. In cases like this bundle utilities can be a big help. But they are not trivial to use. And actually, I am not even sure if you use cmake as build system?

**Comment by stekyne** [comment]
Yeah we're using CMake as well to build the project. 

I'm not sure I understand what you mean by "copy libraries within the vcpkg system"? I don't think there is a requirement that client projects of vcpkg be themselves libraries of vcpkg? The functionality I'm looking for it present for executable files, dumpbin will find all dependent dll files and attempt to find them in the vcpkg installed dir, if it finds it, it copies it into the output folder. The only problem is that this script doesn't get applied to projects in my solution that output a dll instead of an executable. 

https://github.com/Microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L111
I can see in this code that this seems to be the design, the add_library function (line 122) doesn't seem to have this copy dependencies functionality. Is it possible for it to be added?

The only sane way for this to work is that everything ends up in the output directory, otherwise as you suggested, we'd have to start managing the copying and bookkeeping of the files manually.

**Comment by bagong** [comment]
Makes sense what you're saying. Hopefully somebody who understands this better can answer ;)

**Comment by ras0219-msft** [comment]
Forgive me for being a bit brief (mobile + weekend).

We don't copy dependencies for libraries because they aren't directly executable, so the theory was that whenever you do make the executable (even if in another build) you can use vcpkg to copy the dependencies at that time. As long as _your_ libraries are deployed adjacent to the executable, we should recurse into it and pick up any private deps as well.

Without suggesting that you should restructure your cmake project, there are some cmake globals that can be set to override the default output locations and redirect them to the same folder, even in a build with subdirectories. These globals are used in cpprestsdk to handle the deployment issues of having many DLLs and exes produced by the build.

Now, there are some cases where DLLs are intended to bring their own dependencies. In those circumstances, it would be useful to have some way to opt-in to vcpkg's deployment behavior for those libraries.

**Comment by stekyne** [comment]
Thanks for the response. 

It's still not really 100% clear. Can you elaborate more on the "deployed adjacent to the executable"? As far as I know, that is the case. 

Also, how should the CMake project be structured to best fit vcpkg and what are the globals that can be overridden? 

Thanks

**Comment by ras0219-msft** [comment]
By deployed adjacent to the executable, I mean that any DLLs you build _during_ the build that are needed for your application are placed into the same output folder as the application before it's built. For example:
```cmake
add_executable(main main.cpp)
add_library(foo SHARED foo.cpp)
target_link_libraries(main foo)
```
Both `main.exe` and `foo.dll` will be in the root of your build directory, so when we run our "deployment" script, we can recurse from `main.exe`, into `foo.dll`, and realize that it depends on `boost-system.dll` which needs `zlib.dll` (for example).

What _won't_ work is
```cmake
# /libfoo/CMakeLists.txt
add_library(foo SHARED foo.cpp)

# /CMakeLists.txt
add_subdirectory(libfoo)
add_executable(main main.cpp)
target_link_libraries(main foo)
```
This doesn't work for two reasons: 1. We can't find `foo.dll`, so we can't deploy its dependencies, and 2. If you try to run `./main.exe`, the Windows loader won't find `foo.dll` either!

One way (probably the easiest way) to fix this is to redirect all output files to the same directory:
```cmake
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
set(CMAKE_PDB_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
```
If this is set in the root CMakeLists.txt, all subdirectory libraries/executables will be put into the root directory. It also has the added advantage that your application is immediately runnable in the output directory 😄.

[1] https://cmake.org/cmake/help/v3.0/variable/CMAKE_RUNTIME_OUTPUT_DIRECTORY.html

**Comment by stekyne** [comment]
Hmm ok, I think because a lot of the projects in the solution are emitting run-time loaded plugin type DLLs, they're not getting recurred into by the applocal.ps1 script. 

> Now, there are some cases where DLLs are intended to bring their own dependencies. In those circumstances, it would be useful to have some way to opt-in to vcpkg's deployment behavior for those libraries.

Is this behaviour that I'm looking for what you alluded to here?

**Comment by ras0219-msft** [comment]
Yeah, that's exactly the case I meant.

However, plugin architectures come with their own host of headaches. I think you will probably be happiest in the general plugin case if you instead static link all your dependencies. This will enable your plugin to use different versions of libraries than the host program or any other plugins -- otherwise you can end up with really random runtime crashes because you accidentally used someone else's binary-incompatible version.

Right now, the best information on how to static link with Vcpkg is via (our blog post)[https://blogs.msdn.microsoft.com/vcblog/2016/11/01/vcpkg-updates-static-linking-is-now-available/], but I hope to fix this very soon with #1111!

Note that by default, x86-windows-static will statically link against the CRT/STL as well, which might not be desirable depending on the exact structure of your plugin framework (plugins are actually really hard to get right!). You might want to create a new triplet for static library linkage + dynamic CRT linkage. more information on that is available in https://github.com/Microsoft/vcpkg/issues/1108#issuecomment-302603649.
